x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/user_selector.tsx

```diff
@@ -0,0 +1,159 @@
+  const usernameOptions = elasticsearchUsers.map(({ username }) => ({
+    id: username,
+    text: username,
+  }));
+  const hasElasticsearchUsers = elasticsearchUsers.length > 0;
+  const showNewUserExistingUserControls = userFormUserIsExisting && hasElasticsearchUsers;
+
+  const roleSelect = (
+    <EuiFormRow label={ROLE_LABEL}>
+      <EuiSelect
+        name={ROLE_LABEL}
+        data-test-subj="RoleSelect"
+        options={roleOptions as EuiSelectOption[]}
+        value={roleType as string}
+        onChange={(e) => handleRoleChange(e.target.value as SharedRole)}
+      />
+    </EuiFormRow>
+  );
+
+  const emailInput = (
+    <EuiFormRow label={EMAIL_LABEL} helpText={!elasticsearchUser.email && REQUIRED_LABEL}>
+      <EuiFieldText
+        name={EMAIL_LABEL}
+        data-test-subj="EmailInput"
+        value={elasticsearchUser.email as string}
+        onChange={(e) => setElasticsearchEmailValue(e.target.value)}
+      />
+    </EuiFormRow>
+  );
+
+  const usernameAndEmailControls = (
+    <>
+      <EuiFormRow label={USERNAME_LABEL} helpText={!elasticsearchUser.username && REQUIRED_LABEL}>
+        <EuiFieldText
+          name={USERNAME_LABEL}
+          data-test-subj="UsernameInput"
+          value={elasticsearchUser.username}
+          onChange={(e) => setElasticsearchUsernameValue(e.target.value)}
+        />
+      </EuiFormRow>
+      {elasticsearchUser.email !== null && emailInput}
+      {roleSelect}
+    </>
+  );
+
+  const existingUserControls = (
+    <>
+      <EuiSpacer size="s" />
+      <EuiFormRow label={USERNAME_LABEL}>
+        <EuiSelect
+          name="Username select"
+          data-test-subj="UsernameSelect"
+          options={usernameOptions}
+          value={elasticsearchUser.username}
+          disabled={!hasElasticsearchUsers}
+          onChange={(e) => handleUsernameSelectChange(e.target.value)}
+        />
+      </EuiFormRow>
+      {roleSelect}
+    </>
+  );
+
+  const newUserControls = (
+    <>
+      <EuiSpacer size="s" />
+      {usernameAndEmailControls}
+    </>
+  );
+
+  const createUserControls = (
+    <>
+      <EuiFormRow helpText={!hasElasticsearchUsers && USERNAME_NO_USERS_TEXT}>
+        <EuiRadio
+          id="existingUser"
+          data-test-subj="ExistingUserRadio"
+          label={EXISTING_USER_LABEL}
+          checked={userFormUserIsExisting && hasElasticsearchUsers}
+          onChange={() => setUserExisting(true)}
+          disabled={!hasElasticsearchUsers}
+        />
+      </EuiFormRow>
+
+      {showNewUserExistingUserControls && existingUserControls}
+      <EuiSpacer />
+      <EuiRadio
+        id="newUser"
+        data-test-subj="NewUserRadio"
+        label={NEW_USER_LABEL}
+        checked={!userFormUserIsExisting || !hasElasticsearchUsers}
+        onChange={() => setUserExisting(false)}
+      />
+      {!showNewUserExistingUserControls && newUserControls}
+    </>
+  );
+
+  return isNewUser ? createUserControls : usernameAndEmailControls;
```

Comment:
[not a change request, general comment for 7.15] In general this file was pretty hard for me to read/parse and visualize what the output would look like - the number of conditionals in here feels like a code smell. In general for the future I'd suggest considering a component architecture like this, e.g.,

```jsx
<NewUserFlyout>
  // heading specific to new users
  <SharedUserComponent />
  // ... etc
</NewUserFlyout>

<ExistingUserFlyout>
  // ... etc.
</ExistingUserFlyout>
```

i.e., sharing low-level components, rather than trying to share 1 high level view/flyout

Reply:
I agree it's smelly for sure. Maybe we could pair on this after FF?

Reply:
Yes, would love to!